streamparse/storm/bolt.py

```diff
@@ -392,50 +385,14 @@ def emit_many(self, tups, **kwargs):
         kwargs['need_task_ids'] = False
         return super(BatchingBolt, self).emit_many(tups, **kwargs)
 
-    def _run(self):
-        """The inside of ``run``'s infinite loop.
-
-        Separated out so it can be properly unit tested.
-        """
-        tup = None
-        try:
-            tup = self.read_tuple()
-            if tup.task == -1 and tup.stream == '__heartbeat':
-                self.send_message({'command': 'sync'})
-            elif tup.component == '__system' and tup.stream == '__tick':
-                frequency = tup.values[0]
-                self.process_tick(frequency)
-            else:
-                group_key = self.group_key(tup)
-                with self._batch_lock:
-                    self._batches[group_key].append(tup)
-        except Exception as e:
-            log.error("Exception in %s.run() while adding %r to batch",
-                      self.__class__.__name__, tup, exc_info=True)
-            self.raise_exception(e)
-
-    def run(self):
-        """Modified and simplified run loop which runs in the main thread since
-        we only need to add tuples to the proper batch for later processing
-        in the _batcher thread.
-        """
-        storm_conf, context = self.read_handshake()
-        self._setup_component(storm_conf, context)
-
-        tup = None
-        self.initialize(storm_conf, context)
-        while True:
-            self._run()
+    def process_tick(self, tick_tup):
+        """Increment tick counter, and call ``process_batch`` for all current
+        batches if tick counter exceeds ``ticks_between_batches``.
 
-    def _batch_entry_run(self):
-        """The inside of ``_batch_entry``'s infinite loop.
-
-        Separated out so it can be properly unit tested.
+        See :class:`streamparse.storm.component.Bolt` for more information.
         """
-        time.sleep(self.secs_between_batches)
-        with self._batch_lock:
-            if not self._batches:
-                return # no tuples to save
+        self._tick_counter += 1
+        if self._tick_counter > self.ticks_between_batches and self._batches:
             for key, batch in iteritems(self._batches):
```

Comment:
This actually breaks the `auto_fail` code in a surprising way.  If a batch fails, and isn't the first batch to be processed, we'll end up failing all batches. If some of those were processed and acked, then it'll throw a Storm error where you're trying to fail tups which were previously acked.  It's not a new bug, I think, but definitely a bug.  The right solution here is to pop off the dict as you process:

```python
for key in keys(self._batches):  # did I python3 right?
    batch = self._batches.pop(key)
    ...  blah blah blah ...
```